aptos-move/move-examples/nft_auction_house/sources/auction_house_own_listing.move

```diff
@@ -0,0 +1,531 @@
+        assert!(table::contains(&mut auctions.all_active_auctions, auction_id), error::not_found(EAUCTION_NOT_EXIST));
+        let auction = table::borrow_mut(&mut auctions.all_active_auctions, auction_id);
+        // min_incremental_price should be bigger than 0, this also ensure the offer_price is bigger than 0
+        assert!(auction.min_incremental_price > 0, error::invalid_argument(EBID_MIN_INCREMENTAL_IS_ZERO));
+
+        // bid increase has to be bigger than smallest incremental
+        assert!(
+            offer_price >= auction.highest_price + auction.min_incremental_price,
+            error::invalid_argument(EBID_INCREASE_TOO_SMALL)
+        );
+
+        // initialize token store when bidding
+        token::initialize_token_store(bidder);
+
+        // get the listing info
+        // auction is still active
+        let now = timestamp::now_seconds();
+        assert!(now <= listing::get_listing_expiration<CoinType>(&auction.listing), error::invalid_argument(EAUCTION_ENDED));
+        // allow participant to withdraw coin 60 secs after auction ends, configurable by each marketplace
+        let bid_id = bid::bid<CoinType>(
+            bidder,
+            token_id,
+            token_amount,
+            offer_price * token_amount,
+            &auction.listing,
+            listing::get_listing_expiration<CoinType>(&auction.listing) + AUCTION_DELAY_TIME_SEC
+        );
+
+        event::emit_event<BidEvent>(
+            &mut auctions.bid_event,
+            BidEvent {
+                market_address: @auction_house,
+                bid_id,
+            },
+        );
+
+        // store the bid for this auction, only higher bid can enter the auction.
+        *(&mut auction.highest_price) = offer_price;
+        *(&mut auction.highest_bid) = bid_id;
+        vector::push_back(&mut auction.bids, bid_id);
+        bid_id
+    }
+
+    /// bidder can cancel their bid before the auction ends to not participate the auction
+    public entry fun cancel_bid<CoinType>(
+        bidder: &signer,
+        auction_id: u64,
+        bid_id_creation_number: u64,
+    ) acquires Auctions {
+
+        let auctions = borrow_global_mut<Auctions<CoinType>>(@auction_house);
+        assert!(table::contains(&auctions.all_active_auctions, auction_id), error::not_found(EAUCTION_NOT_EXIST));
+        let auction = table::borrow_mut(&mut auctions.all_active_auctions, auction_id);
+
+        let now = timestamp::now_seconds();
+        assert!(now < listing::get_listing_expiration<CoinType>(&auction.listing), error::invalid_argument(EAUCTION_ENDED));
+
+        let bidder_address = signer::address_of(bidder);
+        let bid_id = guid::create_id(bidder_address, bid_id_creation_number);
+        let (offer_price, _, _) = bid::get_bid_info<CoinType>(bidder_address, bid_id_creation_number);
+        //  if the bid is highest bid. find next highest bid. Otherwise, direct remove bid from auction
```

Comment:
It seems strange and can open up edge cases to let highest bidder cancel their bid. This doesn't seem typical in the standard auctions?

Reply:
This is product feature. For example, opensea allows you to cancel bid while magic eden doesn't. But generally cancel bid cost gas. I will also collect feedbacks from our devs on this as well